vizualization_app/axis_to_lcd.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * axis_to_lcd.c
+ *
+ *  Created on: 18 дек. 2017 г.
+ *      Author: dkirtoka
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdint.h>
+
+#define AXISDEVNAME argv[1] //"/dev/mpu6050_0"
+#define FBDEV       argv[2] //"/dev/fb1"
+#define FONTNAME    argv[3] //"9x9.fnt"
+
+#define LCD_WIDTH  128
+#define LCD_HEIGHT  64
+
+#define TEXT_MAX 16
+enum {
+	X,
+	Y,
+	Z,
+	AXES_NUM
+};
+
+#define HEAD_AXIS "AXIS"
+#define HEAD_ACCEL "ACCEL"
+#define HEAD_TEMP "TEMP:"
+struct mpu6050_data {
+	int axis[AXES_NUM];
+	int temerature;
+	int accel[AXES_NUM];
+};
+
+struct image {
+	int width;
+	int height;
+	uint8_t *buffer;
+};
+
+
+struct symbol {
+	uint8_t left;
+	uint8_t top;
+	uint8_t width;
+	uint8_t rows;
+	uint8_t buffer[0];
+};
+
+struct font {
+	int interval;
+	int num_chars;
+	int target_height;
+	struct symbol symbol[0];
+};
+
+static struct font *font;
+
+void draw_bitmap( struct symbol *bitmap, int x, int y, int sym_height, struct image *image )
+{
+	int i, j, p, q, x_max, y_max;
+
+	x += bitmap->left;
+	y += sym_height - bitmap->top;
+
+	x_max = x + bitmap->width;
+	y_max = y + bitmap->rows;
+
+	for (i = x, p = 0; i < x_max; ++i, ++p)
+	{
+		for (j = y, q = 0; j < y_max; ++j, ++q)
+		{
+			if (i < 0 || j < 0 ||
+			        i >= image->width || j >= image->height)
+				continue;
+
+			char c = bitmap->buffer[q * sym_height + p];
+			image->buffer[j * image->width + i] = c;
+		}
+	}
+}
+
+static int get_symbol_offset(char c)
+{
+	return (c < '!' || c > 127) ? 0 : c - '!';
```

Comment:
You hardcode characters to bitmaps mapping here,
but use free order (defined by `ascii.txt`) when bitmaps are generated.
This is errors prone - either symbols' table should be embed into generated font file,
or it should be hardcodded for `crfont.c` as well (in common header).